File: impl/app/android/app/src/main/java/com/ak/contexta/domain/repository/VocabularyRepository.kt
package com.ak.contexta.domain.repository

import com.ak.contexta.domain.model.VocabWord
import kotlinx.coroutines.flow.Flow

interface VocabularyRepository {
    /** Observe active (non-deleted, non-mastered) vocabulary entries */
    fun observeActive(): Flow<List<VocabWord>>

    /** Get active entries count */
    suspend fun getActiveCount(): Int

    /** One-shot: get all active vocab words mapped to domain models */
    suspend fun getActiveWords(): List<VocabWord>

    /** Add word to vocabulary (creates new instance).
     * @return new entry ID, or null if already active in vocabulary */
    suspend fun addWord(wordId: Long): Long?

    /** Mark word as "known" — increment streak, auto-master if streak reaches threshold */
    suspend fun markCorrect(entryId: Long, masteryThreshold: Int = 1)

    /** Mark word as "not known" — reset streak */
    suspend fun markIncorrect(entryId: Long)

    /** Remove word from vocabulary (soft delete) */
    suspend fun removeWord(entryId: Long, reason: String = "MANUAL_REMOVAL")

    /** Count distinct words ever added */
    suspend fun countDistinctWords(): Int
}

File: impl/app/flutter/lib/domain/repository/vocabulary_repository.dart
import '../model/vocab_word.dart';

/// 生词本仓储接口（对齐 Kotlin VocabularyRepository.kt）。
abstract interface class VocabularyRepository {
  /// 观察活跃（未删除、未 MASTERED）生词条目。
  Stream<List<VocabWord>> observeActive();

  /// 活跃条目数。
  Future<int> getActiveCount();

  /// 一次性读取全部活跃生词。
  Future<List<VocabWord>> getActiveWords();

  /// 将词加入生词本（新建条目）。
  /// 返回新条目 id；已在生词本中返回 null。
  Future<int?> addWord(int wordId);

  /// 标记认识：streak+1，达到 [masteryThreshold] 自动 MASTERED。
  Future<void> markCorrect(int entryId, {int masteryThreshold = 1});

  /// 标记不认识：重置 streak。
  Future<void> markIncorrect(int entryId);

  /// 从生词本移除（软删除，记录原因）。
  Future<void> removeWord(int entryId, {String reason = 'MANUAL_REMOVAL'});

  /// 统计加入过生词本的不同词数（未删除）。
  Future<int> countDistinctWords();
}
